### PR TITLE
fix weather time + add limits + add reset button + prevent flickering with fade-timeout

### PR DESCRIPTION
**Pull request by alisa911**

### CODE REVIEW

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2112

File: map/src/drawer/components/weather/Weather.js [hunk 1]
```diff
@@ -9,8 +9,8 @@ import WeatherManager from '../../../context/WeatherManager';
 export default function Weather() {
     const ctx = useContext(AppContext);
 
-    const GFS_WEATHER_TYPE = 'gfs';
-    const ECWMF_WEATHER_TYPE = 'ecmwf';
+    const GFS_WEATHER_TYPE = 'gfs'; //step 1 hour, after 24 hours after the current time - 3 hours
```
Comment (RZR-UA) [inline map/src/drawer/components/weather/Weather.js:12]: fix //step -> // step

File: map/src/drawer/components/weather/Weather.js [hunk 2]
```diff
@@ -107,25 +107,49 @@ export default function Weather() {
...
     function getStep(increment) {
-        let step = 1;
-        const time = ctx.weatherDate.getUTCHours();
+        let step;
+        const utcHours = ctx.weatherDate.getUTCHours();
+        step = getGFSStep(utcHours);
+        if (!step) {
+            step = getECWMFStep(utcHours);
+        }
+        if (!step) {
```
Comment (RZR-UA) [inline map/src/drawer/components/weather/Weather.js:116]: add return 0 into this if

**Review by RZR-UA — APPROVED** [review]

**Comment by RZR-UA** [comment]
- [ ] bug with extra-step when switching between weather type

**Review by RZR-UA — APPROVED** [review]

File: map/src/drawer/components/weather/Weather.js [hunk 3]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
         return (item.key === 'wind' || item.key === 'cloud') && ctx.weatherType === ECWMF_WEATHER_TYPE;
     }
 
-    function getStep(increment) {
-        let step = 1;
-        const time = ctx.weatherDate.getUTCHours();
+    function getDesiredStep() {
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:124]: rename getBaseStep

File: map/src/drawer/components/weather/Weather.js [hunk 4]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+        if (ctx.weatherType === ECWMF_WEATHER_TYPE && Math.abs(diffHours) + new Date().getUTCHours() >= 120) {
+            return 6;
+        }
+        if (ctx.weatherType === GFS_WEATHER_TYPE && Math.abs(diffHours) >= 24) {
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:128]: 2 weather types - 2 main loops

File: map/src/drawer/components/weather/Weather.js [hunk 5]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+        }
+        if (ctx.weatherType === GFS_WEATHER_TYPE) {
+            return 1;
+        }
+        return 0;
+    }
+
+    // align-backward (<0) align-forward (>0) else just align if needed
+    function alignedStep(direction, date = ctx.weatherDate) {
+        const desired = getDesiredStep();
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:142]: const baseStep

File: map/src/drawer/components/weather/Weather.js [hunk 6]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+
+        const ideal = direction < 0 ? -desired : direction > 0 ? +desired : 0;
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:144]: const baseStepWithDirection

File: map/src/drawer/components/weather/Weather.js [hunk 7]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+        const idealHours = new Date(date.getTime() + ideal * 3600 * 1000).getUTCHours();
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:145]: const newHoursUtc

File: map/src/drawer/components/weather/Weather.js [hunk 8]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+
+        if (idealHours % desired === 0) {
+            // console.log('ideal', ideal);
+            return ideal;
         }
-        return increment ? step : -step;
+
+        const currentHours = date.getUTCHours();
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:152]: const currentHoursUtc

File: map/src/drawer/components/weather/Weather.js [hunk 9]
```diff
@@ -106,26 +121,38 @@ export default function Weather() {
...
+        const aligned = direction < 0 ? -(currentHours % desired) : +(desired - (currentHours % desired));
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:153]: const alignedStep = currentHours % desired

const completeAlign

File: map/src/drawer/components/weather/Weather.js [hunk 10]
```diff
@@ -29,17 +30,29 @@ export default function Weather() {
     };
 
     function addWeatherHours(ctx, hours) {
-        let dt = new Date(ctx.weatherDate.getTime() + hours * 60 * 60 * 1000);
+        const dt = new Date(ctx.weatherDate.getTime() + hours * 60 * 60 * 1000);
         ctx.setWeatherDate(dt);
     }
 
+    function resetWeatherDate() {
+        const align = alignedStep(0, new Date());
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:38]: const step

File: map/src/drawer/components/weather/Weather.js [hunk 11]
```diff
@@ -29,17 +30,29 @@ export default function Weather() {
...
+        if (align) {
+            const aligned = new Date(new Date().getTime() + align * 60 * 60 * 1000);
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:40]: const alignedTime

File: map/src/drawer/components/weather/Weather.js [hunk 12]
```diff
@@ -29,17 +30,29 @@ export default function Weather() {
...
+
     useEffect(() => {
         if (ctx.currentObjectType === ctx.OBJECT_TYPE_WEATHER) {
             WeatherManager.displayWeatherForecast(ctx, ctx.setWeatherPoint, ctx.weatherType).then();
         }
-        if (ctx.weatherType === ECWMF_WEATHER_TYPE) {
-            if (ctx.weatherDate.getHours() % 3 !== 0 || ctx.weatherDate.getHours() !== 0) {
-                addWeatherHours(ctx, getStep(false));
+        if (ctx.weatherType) {
+            const align = alignedStep(0);
```
Comment (alisa911) [inline map/src/drawer/components/weather/Weather.js:52]: const step

**Pull request merged**